### PR TITLE
Add strawman proposal for async iterators

### PR DESCRIPTION
**Pull request by stephentoub**
This is by no means a finished design; rather, it's a starting point for subsequent discussions, design, and prototyping.

Contributes to #43
cc: @MadsTorgersen

### CODE REVIEW

File: proposals/async-streams.md [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+    while (await enumerator.MoveNextAsync())
+    {
+        Use(enumerator.Current);
+    }
+}
+finally { await enumerator.DisposeAsync(); }
+```
+
+Discarded options considered:
+- _`ValueTask<bool> MoveNextAsync(); T current { get; }`_: There's no benefit to using `ValueTask<bool>` instead of `Task<bool>`, as all possible `Task<bool>` values for synchronous completion can be cached, for asynchronous completion a `Task<bool>` is allocated anyway to back the `ValueTask<bool>`, and `ValueTask<bool>` adds additional overheads not present with `Task<bool>` (e.g. it has both a `T` and `Task<T>` field, which means its builder does to, which means it increases the size of state machine types that await it).
```
Comment (jnm2) [inline proposals/async-streams.md]: Typo, `does to` should be `does too`

Comment (stephentoub) [reply proposals/async-streams.md]: Fixed. Thanks.

File: proposals/async-streams.md [hunk 2]
```diff
@@ -0,0 +1,789 @@
...
+{
+    while (await enumerator.MoveNextAsync())
+    {
+       T item = enumerator.Current;
+       ...
+    }
+}
+finally
+{
+    await enumerator.DisposeAsync(); // omitted, along with the try/catch, if the enumerator doesn't expose DisposeAsync
```
Comment (jnm2) [inline proposals/async-streams.md]: `try/catch` should be `try/finally`

File: proposals/async-streams.md [hunk 3]
```diff
@@ -0,0 +1,789 @@
...
```
Comment (jnm2) [inline proposals/async-streams.md]: `try/catch` should be `try/finally`

Comment (stephentoub) [reply proposals/async-streams.md]: Fixed

File: proposals/async-streams.md (hunk 2 above)
Comment (stephentoub) [reply proposals/async-streams.md]: Fixed.

**Comment by jnm2** [comment]
Everything in this proposal is profoundly satisfying.
I can see it either way, but my preference is for WaitForNextAsync/TryGetNext.

**Comment by CyrusNajmabadi** [comment]
Q1: Should DisposeAsync take a cancellation token?

**Pull request merged**